### PR TITLE
Global Styles: Do not sanitize block style variations based on registration status

### LINKED ISSUE #62303: Section styles: editing a block variation in Styles doesn't reflect in the editor
While recording a demo of section styles with 6.6 beta 1, I noticed that while I could change the variation in Styles, the changes would update on the front end of the site but would revert in the editor after hitting save. Here's a video showing front end vs back end:

https://github.com/WordPress/gutenberg/assets/26996883/a68230ef-d48c-4bbe-81cb-f183b1d8abd1

Here's the demo video I was working on for good measure:

https://github.com/WordPress/gutenberg/assets/26996883/0465c244-4498-4c94-bcad-40654bfbc989

Here's the theme.json:

```
	"styles": {
		"blocks": {
			"variations": {
				"Dark": {
					"blockTypes": [
						"core/group",
						"core/columns"
					],
					"color": {
						"background": "#1e1e1e",
						"text": "#FFF"
					},
					"elements": {
						"heading": {
							"color": {
								"text": "#FFF"
							}
						}
					}
				}
			}
		}
	}
```

I am assuming this is likely because I'm _adding_ an element style to the variation (button in this case) but that doesn't explain why the Heading doesn't update. This could very well be user error but wanted to report just in case cc @aaronrobertshaw

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Fixes: https://github.com/WordPress/gutenberg/issues/62303

## What?

Only sanitize the properties of block style variations within theme.json instead of also omitting any that are not currently registered.

## Why?

With the introduction of the Section Styles feature that leverages extended block style variations, there are a number of methods through which block style variations can be registered. This primarily happens via the theme.json data filters in the `WP_Theme_JSON_Resolver` class.

Unfortunately, there are several places where global styles data is processed (such as the global styles REST API endpoint) that do not apply these filters and instantiate the `WP_Theme_JSON` class directly. The result then is some block style variations are not registered and are incorrectly stripped during the sanitization of the theme.json data.

## How?

This PR:
- Removes the sanitization of unregistered block style variations
- Maintains the sanitization of style properties within all block style variations
- Prevents any unregistered block style variations from being included in a block types node data during style generation
    - This avoids issues where a variation does not have a defined selector and invalid CSS is generated

## Testing Instructions

There are some [duotone support actions](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/duotone.php#L21-L22) in Gutenberg that trigger all the theme json filters. So to test this fix fully, those actions need to be temporarily disabled.

1. Comment out the [duotone support actions](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/duotone.php#L21-L22) that masked this issue
2. Add a custom block style variation to your theme.json (see https://github.com/WordPress/gutenberg/issues/62303 for snippet)
3. Navigate to Appearance > Editor
4. Select a group block and apply the custom variation
5. Navigate to Styles > Blocks > Group > dark (variation name)
6. Change the background color for the variation and confirm it is reflected in the editor
7. Save the changes and confirm the background color is still correct in the editor
8. For good measure check the frontend
9. Uncomment the duotone support actions from step 1 and repeat the process. It should still work.

## Screenshots or screencast 

| Before | After |
|---|---|
| <video src="https://github.com/WordPress/gutenberg/assets/60436221/a7d59fad-29a1-40d1-84c7-9606f170aef5" /> | <video src="https://github.com/WordPress/gutenberg/assets/60436221/3ad6e58b-d05b-4048-8e35-79987ed1808c" /> |

### CODE REVIEW

File: phpunit/class-wp-theme-json-test.php [hunk 1]
```diff
@@ -1463,12 +1463,14 @@ public function test_get_stylesheet_generates_fluid_typography_values() {
...
-			$theme_json->get_stylesheet()
+			$stylesheet
```
Comment (aaronrobertshaw) [inline phpunit/class-wp-theme-json-test.php:1473]: It turned out the deregistration of the test block before generating the stylesheet was causing this test to fail on this branch. Not 100% sure why yet but I'll look into that on Monday.

**Comment by oandregal** [comment]
Hi Aaron, after your braindump (thank you!) I took a look at this issue and realized there is an alternative way to fix it: we can register the block style variations defined by the theme on `init` instead of using `wp_theme_json_data_*` filters.

The `init` hook is what themes had to use until now in their `functions.php`. Unless I'm missing some use case, this is a safe place to register the block style variations as we can always count on them being registered. I think it's best if we do that rather than trying to patch specific paths, sanitization, etc. It's also simpler to reason about.

The proposal is at https://github.com/WordPress/wordpress-develop/pull/6756 I had some issues in my environment and ended up preparing it for core.

File: lib/class-wp-theme-json-gutenberg.php [hunk 2]
```diff
@@ -2632,10 +2626,14 @@ private static function get_block_nodes( $theme_json, $selectors = array() ) {
 			$variation_selectors = array();
 			if ( isset( $node['variations'] ) ) {
 				foreach ( $node['variations'] as $variation => $node ) {
-					$variation_selectors[] = array(
-						'path'     => array( 'styles', 'blocks', $name, 'variations', $variation ),
-						'selector' => $selectors[ $name ]['styleVariations'][ $variation ],
-					);
+					// Only add registered variations to this node's variations to
+					// avoid CSS generated without a selector.
+					if ( isset( $selectors[ $name ]['styleVariations'][ $variation ] ) ) {
```
Comment (oandregal) [inline lib/class-wp-theme-json-gutenberg.php:2631]: So, what this PR does is:

- do not remove the **unregistered** variations when we build the theme.json in memory
- but make sure we only process the **registered** ones when we use them

This is the kind of thing that I mentioned: we are not removing sanitization but moving it **everywhere**. Each time we have to work with them, we need to make sure to use the ones that are registered.

**Comment by aaronrobertshaw** [comment]
Closing this PR in favour of moving the block style registration to `init` as [detailed above](https://github.com/WordPress/gutenberg/pull/62405#issuecomment-2155131064).

**Pull request closed without merging**